src/PolyType/Abstractions/IConstructorShape.cs

```diff
@@ -29,47 +29,56 @@ public interface IConstructorShape
     /// Includes all formal parameters of the underlying constructor,
     /// as well as any member that can be specified in a member initializer expression.
     /// </remarks>
-    IReadOnlyList<IParameterShape> Parameters { get; }
+    public abstract IReadOnlyList<IParameterShape> Parameters { get; }
+
+    /// <inheritdoc cref="DeclaringType"/>
+    protected abstract IObjectTypeShape DeclaringTypeNonGeneric { get; }
 
     /// <summary>
     /// Accepts an <see cref="TypeShapeVisitor"/> for strongly-typed traversal.
     /// </summary>
     /// <param name="visitor">The visitor to accept.</param>
     /// <param name="state">The state parameter to pass to the underlying visitor.</param>
     /// <returns>The <see cref="object"/> result returned by the visitor.</returns>
-    object? Accept(TypeShapeVisitor visitor, object? state = null);
+    public abstract object? Accept(TypeShapeVisitor visitor, object? state = null);
 }
 
 /// <summary>
 /// Provides a strongly typed shape model for a given .NET constructor.
 /// </summary>
 /// <typeparam name="TDeclaringType">The declaring type of the underlying constructor.</typeparam>
 /// <typeparam name="TArgumentState">The state type used for aggregating constructor arguments.</typeparam>
-public interface IConstructorShape<TDeclaringType, TArgumentState> : IConstructorShape
+public abstract class IConstructorShape<TDeclaringType, TArgumentState>(ITypeShapeProvider provider) : IConstructorShape
 {
     /// <summary>
     /// Gets the shape of the declaring type for the constructor.
     /// </summary>
-    new IObjectTypeShape<TDeclaringType> DeclaringType { get; }
+    public new virtual IObjectTypeShape<TDeclaringType> DeclaringType => (IObjectTypeShape<TDeclaringType>)provider.Resolve<TDeclaringType>();
+
+    /// <inheritdoc/>
+    protected override IObjectTypeShape DeclaringTypeNonGeneric => this.DeclaringType;
+
+    /// <inheritdoc/>
+    public override object? Accept(TypeShapeVisitor visitor, object? state = null) => visitor.VisitConstructor(this, state);
```

Comment:
Do we want to mark this sealed?
```suggestion
    public sealed override object? Accept(TypeShapeVisitor visitor, object? state = null) => visitor.VisitConstructor(this, state);
```

Reply:
In general I rarely apply `sealed` to anything (class or member) and think it's overused as it ties the hands of people like myself that want to do interesting thing later. But I'm perfectly willing to add it if _you_ want me to. I doubt there's a scenario for further overriding this method, so `sealed` seems harmless in this case.

Reply:
Change applied to Accept and Invoke methods everywhere.